OTEP: Define EntityProvider

tedsuo opened the pull request:
This OTEP defines a EntityProvider, to enable the updating of resources that have a lifespan shorter than the lifespan of the application instance.

This OTEP is currently a draft. The primary open question is how to handle active spans that are already running when a resource change is made. Please see the OTEP for details.

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+of the application. And multiple sessions may start and end over the lifespan of
+an application.
+
+Therefore there is no need to conflate "identifying" with "immutable." Telemetry
```
+1 for writing this down. I think this was the biggest insight in the Entities SIG

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+A ResourceListener is a function that takes a resource reference as a parameter.
+Resource listeners SHOULD NOT be required to be thread safe.
+
+### ResourceProvider
```
A few questions on this in context of [OTEP 264](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/entities/0264-resource-and-entities.md)

- Should you be able to add/remove discovered `entity` here or just merge in a resource wholesale as the only capability?
- Should there be an option where pulling resource is a "callback" and ResourceProvider can dynamically discover + cache resource on demand? (i.e. instead of push-based, it's pull based)?  
- Should you be able to "schedule" entity/resource detection to re-occur at regular intervals or via a call? (e.g. similar to `PeriodicMetricReader`, we provide a mechanism to automatically recalculate resource at intervals.

*Note: just call this out in alternatives considered with rationale, if rejecting*

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+## Open questions
+
+The primary open question – which must be resolved before this OTEP is accepted –
```
I assume pending acceptance of this OTEP we'd begin experimenting here?

This seems like the biggest point to address ASAP.

tedsuo replied:
We need to prototype extensively before accepting this OTEP. Like all OTEPs, it should not be accepted until we can link to working examples that illustrate how we plan to resolve these concerns.

bidetofevil commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+as search indexes and metric dimensions. For those situations, we only get to pick
+one value.
+
+The simplest implementation is for the BatchProcessor to listen for resource changes,
```
The naive implementation could lead to a degenerate case, for example, on mobile, where something like a network connection quickly oscillates between connected and unconnected. This would effectively nerf the batching, cutting a new batch a few times a second (which is big in mobile). 

Whether this case needs to be handled well, or simply called out so that implementations can protect itself from it, is something we should address. I don't think a that we need so solve this well at this point for us to move this forward.

jsuereth replied:
SO regarding flagellations + descriptive attributes - I think we need to encourage (but not force) users not to have descriptive attributes in RESOURCE, but allow them in ENTITY.  Additionally, we should prefer NOT to identify source of telemetry with something that is volatile.  These concerns should be addressed in how we model entity types within Semantic Conventions.

We discussed this a bit in person, but effectively things that MAY be stable in a server-side context MAY NOT be stable in a mobile context (or less stable).  So choice of which "entities" to use may need to be localized to an application/service deployment.  OTEL needs to be flexible enough to support these both. What we should NOT do is *prevent* valid identities on Mobile because they don't work in servers or vice-versa.

tedsuo replied:
Yes, but let's be careful to not overcomplicate things with a lot of philosophy! We run into a lot of trouble and bike shedding when we lose track of the practical usage of this data.

Resources are labels applied to all of the data in a batch. Entities are groups of labels. EntityStates are a timeline of changes with precise timestamps as to when the change occurred.

From that perspective, we can ask the question: if a resource such as the networking stack thrashes, or even changes once in the middle of an operation, what label (or labels) is the most helpful to put on that batch of data? And at what granularity should we be segmenting batches of data? The fine grained history is in the EntityState stream. Anything in resources is going to be very coarse grained. So we need to pick a course-graining strategy, based on how we intend backends to make use of the resource labels as a practical manner – not based on a philosophical theory.

tedsuo replied:
For example, perhaps some entities can be marked as "volatile" in a way that causes them to stack up in the same batch rather than trigger a flush.

MrAlias commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+## Trade-offs and mitigations
+
+This change should be fully backwards compatible, with one potential exception:
+fingerprinting. It is possible that an analysis tool which accepts OTLP may identify
+individual services by creating an identifier by hashing all of the resource attributes.
```
Does this mean this will be in a `v2` of the SDK given this is backwards incompatible?

tedsuo replied:
Yes I believe it would, especially if all of the details about provider setup are exposed (which they are in v1). In a world where users have a config file and a NewSDK constructor that encapsulates all of these details, I think it would break far fewer users. But it would still be a major version bump, I don't see how it couldn't be.

In general, it seems like switching to entities would change how resources are detected, and I imagine that alone would probably create a breaking change to SDK setup.

jsuereth replied:
We had been good ng through contortions to try to avoid a bump.

If we think it's easier to just bump the sdk version - we need to discuss that with sdk maintainers but it *would* give us a lot of options in API design.

tedsuo replied:
I think that Go may be in a special situation here.

Breaking the SDK is not great, it would be better if there is overlap and a grace period rather than a hard break.

For reference, this is the [upgrade strategy](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/upgrading.md) we use, which depends on users being able to push to the latest version of the SDK. We don't want to create a situation where maintainers have to maintain two branches.

tigrannajaryan commented:
I support the idea of allowing the Resources to be changed over time. We still need to make sure this complies with our stability guarantees and [Resource spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md).

We will need to specify how this will interact with the entities. The Entities SIG has a proposal about updating the Resource in a specific way and I want to make sure we don't introduce conflicting ideas that will be impossible to merge in the future.

feldentm-SAP commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+// Whenever the SessionManager starts a new session
+// it updates the ResourceProvider with a new session id.
+sessionManager.OnChange(
```
here, the OnChange function is provided by some third-party class and the SetAttribute is used to propagate changes. Is this intended?

feldentm-SAP commented on oteps/4316-resource-provider.md:
````diff
@@ -0,0 +1,272 @@
+  }
+);
+
+```
````
I'd use a before / after resource and have the code in between to make the example more digestable.

feldentm-SAP commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+## Example Implementation
+
+Pseudocode examples for a possible Validator and ResourceProvider implementation. Attention is placed on making the ResourceProvider thread safe, without introducing any locking or synchronization overhead to `GetResource`, which is the only ResourceProvider method on the hot path for OpenTelemetry instrumentation.
```
why is locking relevant here? it seems to be the current focus topic; I agree that it is important, but it isn't the most important topic, right? Using the API and migration should be the most important topics.

feldentm-SAP commented on oteps/4316-resource-provider.md:
````diff
@@ -0,0 +1,272 @@
+
+```
+// Example of a thread-safe ResourceProvider
+class ResourceProvider{
````
I do not really understand the pseudo code syntax. It seems to be Go with classes and some strange extra syntax that I do not get. Also, explicit this is really uncommon and only used in languages that require it due to bad language design.

jsuereth replied:
I'll make this comment again - The syntax/language for pseudo-code isn't important here.

Are you able to understand what the goal of the interface is from the description and the example?  If so, let's evaluate that, not choice of pseudo-code syntax.

If you have specific things you don't understand, list them so they can be addressed.

tedsuo replied:
I would add that regardless of the pseudo-code someone chooses to use, we request that the examples to be 100% explicit and not assume that the reader knows any implicit details about a particular programming language.

feldentm-SAP commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+    // calling listeners inside of the lock ensures that the listeners do not fire
+    // out of order or get called simultaneously by multiple threads, but would
+    // also allow a poorly implemented listener to block the ResourceProvider.
```
Yes, plus the statement above is not really correct since the listener could just enqueue a task in a thread pool. I'm not sure why this part should be specified here. It isn't even required in languages/runtimes that do not have threads or commonly don't use them.

tedsuo replied:
Yes when it comes to language-specific examples it's best to look at prototypes, not pseudo-code. The pseudo-code is helpful to show that there is at least one way to implement the proposed changed in a multi-threaded language. It shouldn't be considered the only way or even the best way, but just something to allow us to discuss potential issues with the design that implementations may need to think about.

jsuereth replied:
Going to mark this as closed, as I think we should rely on the prototypes over the pseudo-code in this design.

feldentm-SAP reviewed (changes requested):
Overall, I appreciate the document. I left picky comments to allow following my perception when reading it top-down without knowing more details.
Things that need to be changed:
 - structure and links to other documents need to be improved before releasing the document
 - the document should provide a clear context at the top and stay in that context
 - we might have to define elsewhere where which kinds of changes are expected
 - I struggle with the pseudo code syntax and the provided signatures seem to be incomplete; maybe adding explicit void could help
 - unless I got something completely wrong, the change is breaking and we should spend more effort on explaining why it is necessary and why it wasn't avoidable

tigrannajaryan commented:
We have this wording in a Stable spec doc:

>When used with distributed tracing, a resource can be associated with the [TracerProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#tracerprovider) when the TracerProvider is created. **That association cannot be changed later.**

How do we reconcile this OTEP with that last assertion? One possibility would be that a Resource directly associated with TracerProvider cannot be changed, but a Resource associated with the ResourceProvider can be replaced and thus the indirect association of the Resource with TracerProvider can change. I would want to confirm that this is an allowed modification of spec wording and we are not breaking our stability guarantees.

jzwc commented:
Thanks for the comprehensive proposal.

Let me comment here on the whole topic from a narow position of a mobile, especially real user monitoring and analytics area.

The proposal, in its analysis, mixes, in my opinion, two, potentially several, different things together: the truly immutable resources (at least on mobile - like operating system, its version, hardware, etc.) and the mutable states of the device (connection QoS) and application (background-foreground, session, user, and their attributes).

My opinion is that just making resources mutable does not solve the apparent inadequacy of the open-telemetry model for the RUM and analytics realm.

The mutable state requires different handling than resources, on several levels. Not only does the state mutate, but it can also be different for different concurrently running bits of code. E.g., a background thread indeed is subject to the same mutable device state (online/offline), yet it can still run in a context of some state attributes in which it was spawned (e.g., session), although the respective app state on the main thread has already changed.

Thus, there seem to be at least three different “resources” realms:

the classic immutable resources as defined in OTel (with potential mutability proposed here)

mutable global state that applies to all running code of the application

mutable state that is typically global but can define different (typically temporary) local context for concurrently running code

Thus it would be, in my opinion, beneficial to narrow the “let us make resources mutable“ proposal to the use cases where the seemingly immutable (by current definition) resources may mutate.

As for other use-cases of mutable device and applications state, a dedicated solution would serve them better. The dedicated channel to sent the mutable state attributes, mentioned in a comment, seems as a sound solution for me. For the local “sticky” state attributes, they can perhaps be provided in the affected signal attributes to override the global state values.

jsuereth commented:
Responding to @jzwc 

> Thus, there seem to be at least three different “resources” realms:
>
> 1. the classic immutable resources as defined in OTel (with potential mutability proposed here)
> 2. mutable global state that applies to all running code of the application
> 3. mutable state that is typically global but can define different (typically temporary) local context for concurrently running code

I'm not sure if you've seen the OTEP where we split apart Resource, but imagine Resource now as a composable set of Entities (bundles).  Resource represents the context in which telemetry was generated, but I think the thing we learned from Mobile is that the context is not the SDK itself but something more dynamic, and we're trying to respond to that here.  IMO - This proposal is about your (3) - mutable state that is typically global but can define different (typically temporary) local context for concurrently running code.

Imagine that resource is composed of a set of bundles (entities).  90% of the entities are static for the lifetime of the SDK, e.g. OS, hardware, etc.  A few change (like session).  With this proposal *and* the [previous OTEP](mutable state that is typically global but can define different (typically temporary) local context for concurrently running code), you'd just be swapping up the entities that changes wholesale (e.g. Here's a new session, replace the old one and all information about it wholesale).

> As for other use-cases of mutable device and applications state, a dedicated solution would serve them better. 

This is also a thing with the Entities work, there would be a special "Entities" channel that could fire out state change events.

TL;DR; I agree on the surface that if you just viewed this proposal purely from resource perspective, I think we'd wind up mixing concerns as you suggest.  If you layer in the Entities work, I think this should give us the tools we need to solve our problems.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+### ResourceListener
+
+A ResourceListener is a function that takes a resource reference as a parameter.
```
The explanations and details provided can definitely be better, but please have a look at the OTEP template to understand the meaning of each section. 

https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0000-template.md

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+starting and ending without the application being shut down or the browser being
+refreshed).
+
+Tracking these resource changes are critical. Without them, it would be impossible
```
Sorry, but you just said that the example use cases listed were not relevant to you, then suggested an example use case that is not relevant to me. Please be more open minded and consider that these use cases are very relevant to other people in the community, especially the various client implementations, which is where the original motivation for this proposal comes from. 

I'm fine adding additional examples, such as the target_info examples you're suggesting. But I'm not familiar with that particular issue as I do not work with Prometheus. If you want to write it up, I'm happy to include it.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+#### OnChange(resourceListener)
+
+Registers a ResourceListener to be called when MergeResource updates the resource.
```
I've updated the explanation to include a description of loose coupling as we practice it in OpenTelemetry. Let me know if it helps!

jsuereth replied:
Resolving for now, please reopen if still an issue.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+Registers a ResourceListener to be called when MergeResource updates the resource.
+
+#### Implementation Notes
```
Good point, moving this above to the object description.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+The resource reference held by the ResourceProvider SHOULD be updated atomically,
+so that calls to GetResource do not require a lock.
+
+Calls to listeners SHOULD be serialized, to avoid thread safety issues and ensure that
```
Yeah we want to ensure that we understand the basic details of at least one reasonable implementation, otherwise the SDK maintainers who read the spec will be left with a number of open questions.

jsuereth replied:
please reopen if still relevant.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+fingerprinting. It is possible that an analysis tool which accepts OTLP may identify
+individual services by creating an identifier by hashing all of the resource attributes.
+
+In practice, the only implementation we have discovered that does this is the OpenTelemetry
```
Prometheus target_info makes a hash of the entire resource object? It doesn't have a list of specific identifying attributes that it uses?

feldentm-SAP replied:
it uses service name/namespace/instance.id and is, hence, equally not compatible.

tedsuo replied:
I'm not sure I follow. How would adding a transient attribute such as `session.id` or `user.name` to the resource object affect target_info?

feldentm-SAP replied:
The way target_info is currently implemented requires that the set of ressource attributes does not change. If some software relies on that property and you change it, that change breaks that software.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+Beyond fingerprinting, there are no destabilizing changes because the resources
+that we have already declared "immutable" match the lifespan of the application
+and have no reason to be updated. Developers are not going to start messing with
+the service.id arbitrarily just because they can, and resource detectors solve the
```
whoops, that should be `service.instance.id`. It is an example of a stable identifying resource. I will come back and do another pass to add links to this and other places where we reference the spec and semantic conventions.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+Beyond fingerprinting, there are no destabilizing changes because the resources
+that we have already declared "immutable" match the lifespan of the application
+and have no reason to be updated. Developers are not going to start messing with
```
Where these changes break the "immutability" of resources is a question the community continuously asks for clarification on, so it is important that we address it here. But I'm sure that this section could be phrased better.

tedsuo replied:
I agree, but that's actually why using the term "immutable" in the spec was never a very accurate description. It only meant "immutable" from the perspective of the local object in code. Not being specific about how we use that word has definitely led to a lot of confusion.

jsuereth replied:
This all has been rephrased, please reopen if you have concerns.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+An alternative to updating resources would be to create span, metrics, and log
+processors which attach these resource attributes to every instance of every
+span and log.
```
In the future, please read the entire OTEP before commenting. 🙏🙏🙏 

There's a couple of places where you have commented only for the concern to be addressed in the next paragraph. Putting everything in the same paragraph is bad English, short paragraphs with topic sentences are a better structure.

To be clear, in general I do appreciate your comments on the content of this doc, and I appreciate that you are taking the effort to review it!

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+Some resources are used to identify a service instance – `service.name`, `service.instance.id`, etc.
+These resources naturally match the lifespan of the service instance. An "immutability requirement"
+is not necessary in this case because there is no reason to ever update these values.
```
It is not possible because we are recording reality.

If, somehow, the reality was that a service.name for an application changes, then the telemetry must reflect that. I cannot imagine why this would ever occur – it sounds like a terrible program! But it is not OTel's job to dictate to users how their programs are supposed to operate. It is our job to record how they operate.

In practice, that means that resources that identify a service instance do not change during the lifespan of the service instance. End users have absolutely no reason to write code that would arbitrarily change these values.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+is not necessary in this case because there is no reason to ever update these values.
+
+Other resources are used to identify other critical lifespans, and these values
+do change. For example, the user's identity may change as the user logs in and out
```
Sure, I'd agree with that. But I don't think those additional details are very relevant here.

The important point is just to illustrate that the lifespan of the service instance is not the only lifespan that we care about from an observability perspective.

It's important to clarify this because unfortunately, the current specification for resources does make this assumption, and requires resources to be "immutable" for the lifespan of the service instance.

In the past a number of members of the OTel community struggled to understand that this was an error, since these other lifespans are not very common in server-side applications. However, they are very common in client applications. Now that we are pushing to get our client implementations out the door, we need to resolve this issue.

I imagine that there are server-side situations where this happens are well, it would be great to list some of them as well if anyone wants to offer an example.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+batches of telemetry would be poorly indexed. This initial telemetry is critical
+and too valuable to lose due to a late loading of certain resources.
+
+A convenient cudgel with which to beat developers into doing the right thing is
```
That's a core problem that reality hands us unfortunately – we have to spend time gathering these resources as some of them require API calls to various systems, including network calls. This can delay the booting of an application, and possibly block it. But if we don't wait to gather these resources, the initial batch of telemetry will not be correct. Possibly never correct if the resource detector fails. It's an ugly reality with no great solution, only tradeoffs.

At google, they decided that they did not want an application to become available until all of the resources could be gathered, and to fail to launch rather than launch with failed calls to resource detectors. That's a pretty reasonable choice to make; I wouldn't want a complex system degrading into poor observability without intentionally choosing to allow it by switching off a detector or having a timeout.

But, if you want to call the senior engineers at google "people lacking experience in complex software systems" please be my guest, as that is where almost all of these designs related to resources originated. 😄

tedsuo commented:
Ok, big update! I've added Entities to the design. It might even make sense to rename this an EntityProvider instead of a ResourceProvider.

Regardless of the name, I'm positive that this design is not correct, as I'm not even sure which Entity document to be referring to at this point (I used OTEP 0256). But as Cunningham's Law states, "the best way to get the right answer on the internet is not to ask a question; it's to post the wrong answer." 

Please let me know what is missing, I would love to pair with one of the entity sponsors to rewrite this and add further details.

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+correct tools these errors are unlikely, as it is very obvious when these individual
+attributes should and shouldn't change.
+
+### Why were resources immutable in the first place?
```
Yeah, a few things.

From OpenCensus - https://github.com/census-instrumentation/opencensus-specs/tree/master/resource
From Stackdriver (Now Google Cloud Observability) - https://cloud.google.com/monitoring/api/resources

There may be others. I have a write-up (in Entities WG notes document) about lessons learned from Google Cloud's resource design.

tedsuo replied:
Most proprietary APM products have something like this in them, but they are closed source so I can't point at the code.

tedsuo commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+### SDK Changes
+
+NewTraceProvider, NewMetricsProvider, and NewLoggerProvider now take a ResourceProvider
```
We are allowed to issue new major versions of the SDK constructors, but it is best if we can avoid it. Implementations lacking overload are architected to use other forms of extensibility to allow for additional config parameters, so that part is not really a problem. But we have identified that the Go SDK may break for other reasons, and we need to be careful.

If you're asking about Providers in general, I would suggest reading the tracing and logging portion of the spec, it will give you a good idea of the patterns that we use in the SDK design. Looking at metrics is also a good idea, but it's a bit more complicated than the other signals.

tedsuo commented:
Latest update:

* Added an AddEntity method to ResourceProvider
* Clarified that the resource object should be regenerated entirely when an entity changes

The next big bugbear I would like to look at is how to handle the potential for thrashing  resources, such as the networking stack on mobile rapidly changing.

One possible solution is to batch changes in a time window. Rather that have separate individual callbacks for EntityState or EntityDelete, there would be a single callback that receives a list of entity events along with the final computed resource. Adjusting the time window length makes a tradeoff between the maximum number of batch flushes that can be triggered and the accuracy of the resources on any particular batch. I don't know if that is a good solution. @bidetofevil what do you actually want here?

carlosalberto commented:
Removing the stale label (hope there's still intention to work on this @tedsuo )

tedsuo commented:
Coming back to this again. The remaining work on this OTEP related to the potential for thrashing resources. After investigating this issue, it does not appear to be a problem. If it is a problem, it is very rare and existing implementations (such as RUM client) do not address it. I've updated the OTEP to clarify that any throttling that may be needed should be implemented in the specific resource detector, not in the provider itself.

The other discovery was that in Otel Java, we already have an object called a `ResourceProvider`. To prevent a naming conflict, I've renamed this object to be an `EntityProvider`.

Until I receive feedback, I have no more work to do on this draft proposal. Prototyping should probably be the next step.

tedsuo commented:
Resurrecting this now that I'm back. Again, at this stage I consider this proposal to be complete, I have no further changes to propose. We need prototypes and feedback to continue.

breedx-splk commented:
@tedsuo Wondering what the intended behavior is if `Add Entity` is called with an entity whose ID is already in the map? Is it a no-op or is it an update?

Another thought is to consider combining update and add into an upsert method instead.

bidetofevil commented:
> One possible solution is to batch changes in a time window. Rather that have separate individual callbacks for EntityState or EntityDelete, there would be a single callback that receives a list of entity events along with the final computed resource. Adjusting the time window length makes a tradeoff between the maximum number of batch flushes that can be triggered and the accuracy of the resources on any particular batch. I don't know if that is a good solution. @bidetofevil what do you actually want here?

Shoot, I missed this when it was posted.

I don't think managing the backpressure of repeated calls on thrashing resources to do a merge or just throw things away is a big deal, so lets not make that layer more complicated.

What I want is not having to flush a batch just because a resource attribute is thrashing. Perhaps there's a way to represent a frequently changed resource attribute value within one batch of telemetry in a terse way? Or is that a non-starter because it'll put to much burden on the collector to unpack it, or require a much bigger change to the spec?

tedsuo commented:
PR against this unmerged OTEP can be found here:
https://github.com/open-telemetry/opentelemetry-specification/pull/4565

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,403 @@
+* `Resource`: represents the entire set of resources after the entity
+  has been deleted.
+
+### EntityProvider
```
Note: We need to sort out a name for this.

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,403 @@
+* `Update Entity`
+* `Delete Entity`
+* `Get Resource`
+* `On Change`
```
I'm onboard with `on change` but I didn't have a chance to create that in my prototype. I'm still focused on raw CRUD operations and lining up Resource Detection "v2".

Wanted to confirm though, the `On Change` portion of this is meant to be an "SDK communication" feature right?

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,403 @@
+### SDK Changes
+
+NewTraceProvider, NewMetricsProvider, and NewLoggerProvider now take a EntityProvider
+as a parameter. How SDKs handle resource changes is listed under [open questions](#open-questions).
```
Update on this - Need to sort out how MeterProvider responds to changes going forward, but we think we have a path for TracerProvider / LoggerProvider.

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+## Trade-offs and mitigations
+
+This change should be fully backwards compatible, with one potential exception:
```
Agreed on the wording of backwards compatibility.  The SIG discussed this for the past two weeks and I've updated the wording to account for it. PTAL.

jsuereth commented on oteps/4316-resource-provider.md:
```diff
@@ -0,0 +1,272 @@
+
+## Example Usage
+
+Pseudocode example of a ResourceProvider in use. The resource provider is loaded with all available permanent resources, then passed to a TraceProvider. The ResourceProvider is also passed to a session manager, which updates an ephemeral resource in the background.
```
This has been updated

jsuereth commented:
> We have this wording in a [Stable spec doc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#resource-sdk):
> 
> > When used with distributed tracing, a resource can be associated with the [TracerProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#tracerprovider) when the TracerProvider is created. **That association cannot be changed later.**
> 
> How do we reconcile this OTEP with that last assertion? One possibility would be that a Resource directly associated with TracerProvider cannot be changed, but a Resource associated with the ResourceProvider can be replaced and thus the indirect association of the Resource with TracerProvider can change. I would want to confirm that this is an allowed modification of spec wording and we are not breaking our stability guarantees.

This is now addressed in the OTEP.  It no longer states this is non breaking and instead offers a path forward.
